Add deep delete functionality

kbrock opened the pull request:
# links

- [x] https://github.com/ManageIQ/manageiq/pull/22670 (2 commits)
- [x] https://github.com/ManageIQ/manageiq/pull/22669 (2 commits)
- [x] https://github.com/ManageIQ/manageiq/pull/22694
- [x] https://github.com/ManageIQ/manageiq/pull/22695

### Problem

We need to destroy an ems. Unfortunately, the ems has a lot of child objects. Too many objects are loaded into memory and the sheer number of queries means this takes up too much memory and time.

It was traversing each layer; loading all the objects in a layer and then loading all child objects into memory before destroying the layer. This is a typical N+1 `parent_id` algorithm and can result in loading billions of objects into memory.


### Solution approaches

The solution either needs to a) cut up the problem into smaller chunks that avoids loading the entire ems tree into memory, b) avoiding loading relations into memory, c) some combination of those two approaches.

Converting `:dependent => :destroy` to `:dependent => :delete` does this, and can be done if there are not ruby callbacks.

By extension, if an object has no ruby callbacks and only has `:dependent => :delete`, then these can records can be deleted (not destroyed), just so long that the same code deletes the children first.

As you can guess, you end up with a ton of `DELETE` queries with a root of `WHERE ems_id = $ems_id`.
This reduces the N+1 queries, and no longer loads the leaf records or tree records into memory.

For records that have a callback defined, they are brought into memory and destroyed the old fashion way. But most `:dependent => :destroy` are converted to `:dependent => :delete`. And these deletes are done as a single query from the top level.

### Before

Each object is traversed and all children are then destroyed.
Each of those children's children are then destroyed before the child can be destroyed.

This is a typical N+1 `parent_id` scenario and for a large EMS. It takes 45 minutes on macbook metal.

Issues: 

- Object in each tier needs to be loaded into memory before the children can be brought back and destroyed.
- The records are brought back in N+1 fashion, and deleted individually. There are tens of thousands of containers. Ignoring the child relations for those containers, this overwhelms our system.

### After

#### Algorithm

This algorithm is similar to the original N+1 algorithm but with a few caveats:

- use ruby `has_many` to define children rather than loading the objects into memory (via `refs_callbacks`)
- Instead of locating each child for each object, all children for each layer are referenced  together (via `rel_scope`).
- deepest leaf children are deleted first.
  - If a relation is a `nullify` then the records are all updated with a single query.
  - If a relation is a `delete` or a `destroy` but there are no ruby callbacks, then all children are deleted with a single query
  - If a relation is a `destroy` and it has ruby callback, then `destroy` it. These are loaded into memory

#### Implications

(going through the above list bullet by bullet:

- `has_many`
  - Objects no longer need to be loaded into memory.
  - For STI classes, care needs to be made to ensure all relations are properly discovered since STI child classes can also define relations or alter relations.
- `rel_scope`
  - referencing all children at each level across parents increases our efficient from `O(n)` where n is the number of records to `O(associations)` or around 100.
  - Referencing all children at each level means we need to reimplement their scope. This is error prone.
  - see limitations below
- deepest leaf first
  - traversing a tree means this is still a recursive algorithm.
  - `:nullify`
    - This is a simple `UPDATE` like stock rails. The difference being that it is a single query for all updates.
  - `:delete`
    - These are also not loaded into memory. The code favors `:destroy` over `:delete`, but if there are no callbacks, then there is no need for `:destroy`.
 - `:destroy`
   - If we do destroy, it will attempt to destroy all relations. Since we walk the tree from child to parent, we will hit a number of queries, but they will all bring back no records. I had tried to fix this but this simplicity with a dozen or so no-op queries seemed minor.

### Limitations

2 main limitations:

#### Constraints

When building the child scopes, constraints on the relations are ignored. (this is my limitation) I reviewed all of the constraints and none of them altered the records we wanted to delete. Most of the constraints overlapped with other relations that were the same, just more genthat deleted them. This is the reaosn for the 2 PRs that correct the duplicate `:dependent => :destroy` definitions.

This still works correctly without those PR modifications, since only use constraints to divide relations that are still dependent.

#### Child STI classes

We build the relations from the relation class, so if relations are defined in a STI child class, those relations will not get picked up. This is the reason why the entry `delete` method rebuilds the query to ensure the `class` is not a generic `ExtManagementSystem` but rather something like a `ContainerManager`. This is only an issue if the relation defined specifies a parent class when the actual child class adds relations.

It seems that only the Ems adds relations to classes, so the top level ems is at risk. The ContainerManager does link to other ExtManagementSystem providers, but it properly specifies the specific class, so this works properly. Again, it is only an issue if a relation is defined that points to a generic class that has a child class that defines a `dependent => destroy, nullify, delete` relationships.

### Numbers

For a large ems (acania) this was taking 45 minutes on a mac. It takes much longer in an openshift cluster and is killed due to taking too many gigs of memory.

|       ms |   bytes | objects |query |  qry ms |     rows |` comments`
|          ---:|         ---:|         ---:|  ---:|         ---:|      ---:|  ---
|  3,141,337.7 | 3,429,864,842* | 1,397,325,234 |1,755,195 | 2,178,918.6 |  578,412 | destroy
|  9,361.0 | 61,629,818* | 7,759,264 |  292 | 8,265.7 |       14 |`deep_delete`

\* Memory usage does not reflect 679,643,104 freed objects.
\* Memory usage does not reflect 3,592,320 freed objects.

queries as follows:

count|query
-----|-----
129|COUNT
49 |DELETE
87 |SELECT
11 |COMMIT
11 |BEGIN
5  |UPDATE

It appears the 87 `SELECT` queries are to bring back objects for destroying. 51 of those are from the top level ems destroy. Most if not all of those are already removed from deletes. This is probably why there were only 14 records brought back for `SELECT` clauses. Most of the child records are already deleted before we call destroy on objects. Either that or the instantiations were not showing up for some other reason in instrumentation code.

kbrock commented:
update:
- rebased (updated puma) to fix build
- pulled in 2 new prs (to resolve their cops)
- adding specs

kbrock commented:
update:
- rubocops
- added intermediate variables to explain and removed corresponding comments
update:
- fix errors for last rubocops
-

Fryguy commented on lib/deep_delete.rb:
```diff
@@ -0,0 +1,157 @@
+# currently only tested on ContainerManager / ems
+# @param obj an Ems
```
What does this `@param` refer to?

kbrock commented:
updat:
- pushed missed change for adding the logger

update:
- cops: changed `format`, spacing, ignored `update_all`, removed the last `puts`

Fryguy commented:
Now that I've read the algorithm we _absolutely_ should try to upstream this.  Great work @kbrock - really looking forward to this being merged.

I wonder if there are other uses of this.  One of the reasons we have purging mixin is because we were trying to avoid this _exact_ problem of deleting a top-level object and then all of the cascade deletes taking forever.  But if deep delete is sufficient, perhaps we can put all that back.

kbrock commented:
UNWIP: fixed issue with through associations deleting the relation instead of the link. This has been fixed

kbrock commented:
update:
- hash indentation
- dropped unused `run(:scope)`
- block_given => block nil check

kbrock commented:
update:
- rebase
- removed ENV.fetch (sorry, it was debug code)
- removed unused scope parameter (think this was an old cop)

Double verified that `has_many :model, :through => :model2, :dependent => :destroy` destroys the link.

Fryguy commented on lib/deep_delete.rb:
```diff
@@ -0,0 +1,186 @@
+#   DeepDelete.vacuum # to update indexes
+#
+# currently only tested on ContainerManager / ems
+class DeepDelete
```
My only concern that's left is that this class name is a top-level constant, and I'm concerned we might pull in a gem later or something that also defines this constant (it comes close now to `DeepMerge` for example).

Thinking we should namespace this under ManageIQ, so ManageIQ::DeepDelete

kbrock replied:
I'm thinking that it supports `:dependency => :delete` / destroy.

so deep delete seems to ring something for me.

Fryguy commented on spec/lib/deep_delete_spec.rb:
```diff
@@ -0,0 +1,67 @@
+# default_ansible_object (resource)?
```
not sure what this random comment refers to.

Fryguy commented on spec/lib/deep_delete_spec.rb:
```diff
@@ -0,0 +1,67 @@
+
+  # silence the output
+  before do
+    allow(DeepDelete).to receive(:puts)
+    allow_any_instance_of(DeepDelete).to receive(:puts)
```
What is outputting anything anymore?  I don't see puts calls above so maybe this can be removed?  I am wary about a library that has puts anyway (otherwise it should be in the tools dir).

Fryguy commented on spec/lib/deep_delete_spec.rb:
```diff
@@ -0,0 +1,67 @@
+    expect(subnet_port).to be_deleted
+  end
+
+  def deep_delete(obj)
+    DeepDelete.delete(obj)
```
test-wise I'm not a fan of this helper method since it doesn't really save anything and only creates a visual disconnect to the method-under-test.

Fryguy commented on lib/deep_delete.rb:
```diff
@@ -0,0 +1,186 @@
+    self
+  end
+
+  def self.vacuum
+    ActiveRecord::Base.connection.execute("vacuum analyze")
+    self
+  end
```
Is this being used?

Fryguy commented on lib/deep_delete.rb:
```diff
@@ -0,0 +1,186 @@
+    self
+  end
+
+  ### this is the main method
```
Can this doc string be improved?  It doesn't really tell what it does other than being the expected entrypoint.

kbrock replied:
this is the only public method
It is probably not needed

Fryguy commented on lib/deep_delete.rb:
```diff
@@ -0,0 +1,186 @@
+# currently only tested on ContainerManager / ems
+class DeepDelete
+  include Vmdb::Logging
+  # list of callbacks that we have reviewed and can saftely ignore
```
```suggestion
  # list of callbacks that we have reviewed and can safely ignore
```

kbrock commented:
update:
- removed 2 comments, fixed spelling in anoth
- added module manageiq
- removed spec puts stub and deep_delete
- removed vacuum
- changed comment around sti
- changed comment around batching. in_batches introduced way too many queries
- changed `count` to `exist?`

changes:
- split `run` into `run` and `skip_run`
- added batching by 10,000 records for deletes and destroys (many more queries)
